Export to BGEN v1.2

jigold opened the pull request:
Depends on #1698

cseed commented on python/hail/dataset.py:
```diff
@@ -1375,6 +1375,49 @@ def downsample_variants(self, keep):
     @handle_py4j
     @requireTGenotype
+    def export_bgen(self, output, nbits_per_prob=8):
+        """Export variant dataset as BGEN and SAMPLE file.
```
I'd put v1.2 right up here.

cseed commented on python/hail/dataset.py:
```diff
@@ -1375,6 +1375,49 @@ def downsample_variants(self, keep):
+
+        **Notes**
+
+        Writes out the internal VDS to a BGEN and SAMPLE fileset in the `Oxford spec <http://www.stats.ox.ac.uk/%7Emarchini/software/gwas/file_format.html>`__.
```
"Writes out the dataset to a ..."

cseed commented on python/hail/dataset.py:
```diff
@@ -1375,6 +1375,49 @@ def downsample_variants(self, keep):
+        Writes out the internal VDS to a BGEN and SAMPLE fileset in the `Oxford spec <http://www.stats.ox.ac.uk/%7Emarchini/software/gwas/file_format.html>`__.
+        The BGEN file is `v1.2 <http://www.well.ox.ac.uk/~gav/bgen_format/bgen_format.html>`__ and the genotype probabilities are unphased and diploid.
+        The maximum rounding error is given by :math:`\\frac{1}{2^{B} - 1}` where :math:`B` is specified by the parameter ``nbits_per_prob``. For example, if :math:`B = 8`, the maximum rounding error is 1/255 or 0.0039.
+        Multiallelic variants are allowed.
```
allowed => supported.

cseed commented on python/hail/dataset.py:
```diff
@@ -1375,6 +1375,49 @@ def downsample_variants(self, keep):
+        The sample file has 3 columns:
+
+        - ID_1 and ID_2 are identical and set to the sample ID (``s``).
+        - The third column ("missing") is set to 0 for all samples.
```
The third column is called missing?  Or 0 means missing?  What's the third column normally represent?

cseed commented on src/main/scala/is/hail/io/bgen/BgenLoader.scala:
```diff
@@ -214,11 +214,11 @@ object BgenLoader {
     val isCompressed = compressType != 0
 
-    val version = flags >> 2 & 0xf
+    val version = flags >>> 2 & 0xf
```
I'd put parens around this for clarity.

cseed commented on src/main/scala/is/hail/io/bgen/BgenLoader.scala:
```diff
@@ -214,11 +214,11 @@ object BgenLoader {
       fatal(s"Hail supports BGEN version 1.1 and 1.2, got version 1.$version")
 
-    val hasIds = (flags >> 31 & 1) != 0
+    val hasIds = (flags >>> 31 & 1) != 0
```
Parens here, too.

Personally, I like to shift the 1 up instead of the data down if you're comparing to 0.

cseed commented on src/main/scala/is/hail/io/bgen/BgenRecord.scala:
```diff
@@ -128,14 +128,15 @@ class BgenRecordV12(compressed: Boolean, nSamples: Int, tolerance: Double) exten
     val sampleProbs = new Array[UInt](nGenotypes)
 
-    val totalProbInt = ((1L << nBitsPerProb) - 1).toUInt
+    val totalProbInt = UInt((1L << nBitsPerProb) - 1)
```
I thought we're preferring .toUInt in the v1.2 branch?

cseed commented on src/main/scala/is/hail/io/bgen/BgenWriter.scala:
```diff
@@ -0,0 +1,186 @@
+    byteBuffer ++= samplePloidies
+    byteBuffer += phased
+    byteBuffer += nBitsPerProb.toByte
+    byteBuffer ++= bitPack(sampleProbs.result(), nBitsPerProb)
```
Don't do this.  These functions allocate arrays which you're adding to the byte buffer.  Don't do per-genotype allocation.  They should take the byte buffer and add the elements in directly.

cseed commented on src/main/scala/is/hail/io/bgen/BgenWriter.scala:
```diff
@@ -0,0 +1,186 @@
+
+    val sampleProbs = new ArrayBuilder[UInt]
+    val nGenotypes = triangle(nAlleles)
+    val noCall = Array.fill[UInt](nGenotypes - 1)(UInt(0))
```
Don't need fill since 0 is the default.

cseed commented on src/main/scala/is/hail/io/bgen/BgenWriter.scala:
```diff
@@ -0,0 +1,186 @@
+import is.hail.utils._
+import is.hail.variant.{Genotype, Variant, VariantDataset}
+
+import scala.collection.mutable.ArrayBuffer
```
Unused.

cseed commented on src/main/scala/is/hail/io/bgen/BgenWriter.scala:
```diff
@@ -0,0 +1,186 @@
+  def emitGenotypeData(gs: Iterable[Genotype], samplePloidies: Array[Byte],
+    nAlleles: Int, nBitsPerProb: Int, conversionFactor: Double, bitMask: Long): (Array[Byte], Int, Int) = {
+    val byteBuffer = new ArrayBuilder[Byte]
+    val compresser = new Deflater()
```
compressor?

cseed commented on src/main/scala/is/hail/io/bgen/BgenWriter.scala:
```diff
@@ -0,0 +1,186 @@
+    compresser.setInput(uncompressedData)
+    compresser.finish()
+
+    val outputStream = new ByteArrayOutputStream(uncompressedData.length)
```
I wouldn't involve ByteArrayOutputStream here.  Also, uncompressed length with be unnecessarily big.  You should just grow an array and copy it to make the size right.

cseed commented on src/main/scala/is/hail/utils/ArrayBuilder.scala:
```diff
@@ -36,6 +36,10 @@ class ArrayBuilder[@specialized T](initialCapacity: Int)(implicit tct: ClassTag[
   }
 
+  def ++=(x: Traversable[T]) {
+    x.foreach(t => this.+=(t))
```
I think this shouldn't be allowed.  T will get boxed in Traverable, and this interface is intended to be box-free.

cseed commented:
Just to double-check you're not waiting on me for this one.

jigold commented:
No -- will get to this in the next day or two.

jigold commented on python/hail/dataset.py:
```diff
@@ -1071,6 +1071,51 @@ def sample_variants(self, fraction, seed=1):
+        The sample file has 3 columns:
+
+        - ID_1 and ID_2 are identical and set to the sample ID (``s``).
+        - The third column ("missing") is set to NA for all samples.
```
I double checked the spec. The "missing" column is the missingRate (1 - callRate). NA is the correct value to put if undefined. http://www.stats.ox.ac.uk/%7Emarchini/software/gwas/file_format.html

jigold commented:
@cseed Back to you.

cseed commented on src/main/scala/is/hail/io/bgen/BgenWriter.scala:
```diff
@@ -0,0 +1,179 @@
+    bytes.foreach(bb += _)
+  }
+
+  def bitPack(input: Array[Int], nBitsPerProb: Int): Array[Byte] = {
```
Looks like this is only used in the tests.  I'd move it there.  Looks dangerous, because the block reader should clearly be reusing the ArrayBuilder.

cseed commented on src/main/scala/is/hail/io/bgen/BgenWriter.scala:
```diff
@@ -0,0 +1,179 @@
+        (r._1, math.ceil(r._2))
+      else
+        (r._1, math.floor(r._2))
+    }.sortBy(_._1).map(_._2.toUInt.intRep)
```
OK, this looks right, but the performance is going to be bad.  Basically, don't use any Scala library functions to implement this.  Right now, you're guaranteed input is size 3, right?  Few things:

 - computing totalFractional/F, this is guaranteed to be an integer, so I'd convert it to one.  Since as a floating point value, it might not up exactly to the integer value, round do the nearest integer: `val totalFractional = (sum + 0.5).toInt`.

 - Assert the sum is 2^B-1 at the end.

 - As for the sort, the right way to do this is a function that computes the "sorted index", for each element of an array, it computes the position that element would have in the sorted array.   So if `idx = a.sortedIndex`, then `for (i <- ...) sorted[idx[i]] = a[i]` would be sorted.  I'd allocate the index array on the outside and pass it in so it can be reused.  That way, there are no per-genotype allocations.

 - Finally, to implement `sortedIndex(a, idx)`, input is guaranteed to be length 3, right?  In that case, I might hardcode it.  Otherwise, we should implement it ourself and put it in utils.  This is such a useful function, I don't know why it isn't in the standard library of almost any language.

jigold replied:
Input size can be greater than 3 -- we support multiallelic variants, but UK Biobank will only be biallelic. I'll make two versions -- one general and one hardcode for input size equal to 3.

cseed commented on src/main/scala/is/hail/io/bgen/BgenWriter.scala:
```diff
@@ -0,0 +1,179 @@
+    magicNumbers.foreach(byteBuffer += _)
+    intToBytesLE(byteBuffer, flags)
+
+    byteBuffer.result() ++ sampleIdentifierBlock
```
Oh no!  Don't use ++ on the result, put the sample id block into the byteBuffer (which is a builder).  This will result in a copy of the result of byteBuffer (or worse).

cseed commented on src/main/scala/is/hail/io/bgen/BgenWriter.scala:
```diff
@@ -0,0 +1,179 @@
+    val (gtData, compressedLength, decompressedLength) = emitGenotypeData(gs, nSamples, nAlleles, nBitsPerProb, conversionFactor, bitMask)
+    intToBytesLE(byteBuffer, compressedLength + 4)
+    intToBytesLE(byteBuffer, decompressedLength)
+    gtData.foreach(byteBuffer += _)
```
Have emitGenotypeData take a array builder instead of returning an array.  Basically, for every block of data you're going to emit, you should allocate ONE ArrayBuilder and put everything into it, and only call result() at the very end.

cseed commented on src/main/scala/is/hail/utils/package.scala:
```diff
@@ -473,4 +473,27 @@ package object utils extends Logging
+      val nBytesCompressed = compressor.deflate(buffer)
+
+      while (nBytesCompressed > compressedData.length - compressedOffset) {
+        compressedData = java.util.Arrays.copyOf(compressedData, compressedData.length << 1)
```
nBytesCompressed can't be bigger than the length of buffer (1024), so you only need to do at most one step of this loop.

jigold commented:
@cseed I addressed your comments. I now require the VDS to be split because I hardcoded the sortedIndex to work for the case of 3 alleles. I will keep working on the sortedIndex functionality, but at least this is a working version that could go in independent of that.

cseed commented on src/main/scala/is/hail/io/bgen/BgenWriter.scala:
```diff
@@ -0,0 +1,290 @@
+        idx(2) = 0
+      }
+    } else {
+      throw new UnsupportedOperationException("sortedIndex is only supported for arrays of length 3.")
```
I prefer to to put checks like this at the top.  `if (a.length != 3) throw Blah("foo") ...`.

cseed commented on src/main/scala/is/hail/io/bgen/BgenWriter.scala:
```diff
@@ -0,0 +1,290 @@
+  def sortedIndex(a: Array[Double], idx: Array[Int]) {
+    val n = a.length
+    if (n == 3) {
+      if (a(0) >= a(1) && a(1) >= a(2)) {
```
I think this is more naturally written as a tree: compare `a(0) >= a(1)` and then recurse on the two subcases where 0 appears before 1 and 1 appears before 0.

cseed replied:
I would add a test that compares this to a normal call of sort.

jigold replied:
I wasn't 100% certain what you were suggesting here. Let me know if my interpretation is wrong.

cseed commented on src/main/scala/is/hail/io/bgen/BgenWriter.scala:
```diff
@@ -0,0 +1,290 @@
+    assert(newSize == size)
+  }
+
+  def emitHeaderBlock(vds: VariantDataset): Array[Byte] = {
```
emit sounds like you write the data out.  You're returning it.  I'd simply call it `headerBlock`.

cseed replied:
You could have take the ArrayBuilder, too, but it is only called once (or once per partition once you had `parallel=True`), so I don't really care either way.

cseed commented on src/main/scala/is/hail/io/bgen/BgenWriter.scala:
```diff
@@ -0,0 +1,290 @@
+    bb.result()
+  }
+
+  def emitVariantBlock(v: Variant, va: Annotation, gs: Iterable[Genotype],
```
emit: same.

cseed replied:
Actually, leave the emit, but this should take a ArrayBuilder which is allocated per-partition.  See comment below.

cseed commented on src/main/scala/is/hail/utils/ArrayBuilder.scala:
```diff
@@ -36,6 +40,20 @@ class ArrayBuilder[@specialized T](initialCapacity: Int)(implicit tct: ClassTag[
     size_ += 1
   }
 
+  def ++=(a: Array[T]) {
```
Just call `++=(a, a.length)`.

cseed commented on src/main/scala/is/hail/variant/VariantDataset.scala:
```diff
@@ -235,15 +236,42 @@ class VariantDatasetFunctions(private val vds: VariantSampleMatrix[Genotype]) ex
+      it.map { case (v, (va, gs)) =>
+        BgenWriter.emitVariantBlock(v, va, gs, rsidQuery, varidQuery, nSamples, nBitsPerProb)
+      }
+    }.saveFromByteArrays(path + ".bgen", vds.hc.tmpDir, header = Some(bgenHeader))
```
OK, we need to modify this to support a `export_bgen(..., parallel=True)` option that writes one BGEN file per partition, since the call to `copyMerge` in `saveFromByeArrays` will kill us in creating multiple copies of UKBBv1.

Second, we should create a single ArrayBuilder inside the mapPartitions here and pass it into emitVariantBlock so you don't create an ArrayBuilder for each variant.

cseed commented on src/main/scala/is/hail/io/bgen/BgenWriter.scala:
```diff
@@ -0,0 +1,290 @@
+      } else {
+        gtDataBuffer += ploidy
+        if (gt.isLinearScale)
+          resizeWeights(px, sampleProbs, resized, index, newSize)
```
OK, I'd like to get rid of `sampleProbs`.  I would create a `BitPacker` class that has a pointer to the ArrayBuilder and has `+=(Int)` that you can write the individual probabilities.  This can be passed to `resizeWeights`  which call `BitPacker +=` rather than `sampleProbs +=`.

jigold commented on src/main/scala/is/hail/io/bgen/BgenWriter.scala:
```diff
@@ -0,0 +1,298 @@
+  }
+}
+
+trait IntConsumer {
```
I added this abstract class because I didn't want to duplicate the `roundWithConstantSize` function, but I also wanted to be able to give the function both an Array[Byte] and an Array[Int] for the tests.

jigold commented:
@cseed Back to you. I am planning to make a new PR with functionality to rename files when parallel = true for both exportbgen and exportvcf.
